src/sphinx_antsibull_ext/schemas/ansible_links.py

```diff
@@ -18,13 +18,14 @@ class AnsibleLink(p.BaseModel):
     ref: t.Optional[str] = None
     external: bool = False
 
-    @p.root_validator()
-    # pylint:disable=no-self-argument
-    def one_of_url_and_ref(cls, values: dict[str, t.Any]) -> dict[str, t.Any]:
-        has_url = values.get("url")
-        has_ref = values.get("ref")
-        if has_url == has_ref:
-            raise ValueError("Exactly one of 'url' and 'ref' must be specified.")
+    @p.model_validator(mode="before")
+    @classmethod
```

Comment:
I don't think classmethod is strictly necessary from pydantic's side. Do one of our linters/type checkers complain if it's not there?

Reply:
I don't know whether it's really necessary (I haven't tried), but the (only) example in the pydantic documentation also uses it: https://docs.pydantic.dev/latest/concepts/validators/#model-validators

Reply:
(Also it avoids problems with other linters who don't know that pydantic does this.)

Reply:
> @field_validators are "class methods", so the first argument value they receive is the UserModel class, not an instance of UserModel. We recommend you use the @classmethod decorator on them below the @field_validator decorator to get proper type checking.

I suppose the same thing applies here.

Marking as resolved.